src/System.CommandLine.Hosting/HostingExtensions.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System.CommandLine.Builder;
+using System.CommandLine.Invocation;
+using System.Linq;
+
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace System.CommandLine.Hosting
+{
+    public static class HostingExtensions
+    {
+        public const string ConfigurationDirectiveName = "config";
+
+        public static CommandLineBuilder UseHost(this CommandLineBuilder builder,
+            Func<string[], IHostBuilder> hostBuilderFactory,
+            Action<IHostBuilder> configureHost = null) =>
+            builder.UseMiddleware(async (invocation, next) =>
+            {
+                var argsRemaining = invocation.ParseResult.UnparsedTokens.ToArray();
+                var hostBuilder = hostBuilderFactory?.Invoke(argsRemaining)
+                    ?? new HostBuilder();
+                hostBuilder.Properties[typeof(InvocationContext)] = invocation;
+
+                hostBuilder.ConfigureHostConfiguration(config =>
+                {
+                    config.AddCommandLineDirectives(invocation.ParseResult, ConfigurationDirectiveName);
+                });
+                hostBuilder.ConfigureServices(services =>
+                {
+                    services.AddSingleton(invocation);
+                    services.AddSingleton(invocation.BindingContext);
+                    services.AddSingleton(invocation.Console);
+                    services.AddTransient(_ => invocation.InvocationResult);
+                    services.AddTransient(_ => invocation.ParseResult);
+                });
+                hostBuilder.UseConsoleLifetime();
```

Comment:
@couven92 @jonsequitur it would be nice if we could use the `CancellationToken` from https://github.com/dotnet/command-line-api/blob/a58aa11718049db7a9e49a65cd52db4bfb42e38f/src/System.CommandLine/Invocation/InvocationExtensions.cs#L19 here.

Reply:
@tmds you can do that already. A handler taking both `IHost` and `CancellcationToken` as arugments would still gte the cancellation token from the `BindingContext` as you usually would already.

During the `AddMiddleware` method used in `UseHost` there is unfortunately no way to resolve services added to the `BindingContext`, since I cannot access its `ServiceProvider` property from within the `Hosting` project. And even we could, what would you have me do with the cancellation token? Use in the calls to `host.StartAsync()` and `host.StopAsync`?

I don't think it would be a great idea to add the cancellation token to the host service provider using `ConfigureServices`. If that was your intention, I'd much rather add the `CancellationTokenSource` obtained from the private field `_cts` in `InvocationContext`, but that might not be ideal either.

Reply:
Discussion for this moved to issue https://github.com/dotnet/command-line-api/issues/549.